<?xml version="1.0"?>
<robot name="PF400_Robot" xmlns:xacro="http://www.ros.org/wiki/xacro">


  <!-- Links -->
  <xacro:macro name="pf400" params = "prefix">
  <link name="${prefix}rail_base">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 1.5708"/>
      <geometry>
        <mesh filename="package://pf400_description/meshes/rail_base.STL" scale="1 1 1"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://pf400_description/meshes/rail_base.STL" scale="1 1 1"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 1.5708"/>
    </collision>
  </link>

  <link name="${prefix}rail">
    <visual>
      <origin xyz="0 0 0" rpy="1.5708 0 1.5708"/>
      <geometry>
        <mesh filename="package://pf400_description/meshes/rail.STL" scale="1 1 1"/>
      </geometry>
      <material name="white"/>
    </visual>
      <collision>
      <geometry>
        <mesh filename="package://pf400_description/meshes/rail.STL" scale="1 1 1"/>
      </geometry>
      <origin xyz="0 0 0" rpy="1.5708 0 1.5708"/>
    </collision>
  </link>

  <link name="${prefix}z_column">
    <visual>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Z_Column.STL" scale="0.001 0.001 0.00132"/>
      </geometry>
      <material name="white"/>
    </visual>
      <collision>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Z_Column.STL" scale="0.001 0.001 0.00132"/>
      </geometry>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
    </collision>
  </link>

  <link name="${prefix}z_carriage">
    <visual>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Z_Carriage.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="white"/>
    </visual>
      <collision>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Z_Carriage.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
    </collision>
  </link>

  <link name="${prefix}inner_link">
    <visual>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Inner_Link.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="white"/>
    </visual>
      <collision>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Inner_Link.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
    </collision>
  </link>

  <link name="${prefix}outer_link">
    <visual>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Outer_Link.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="white"/>
    </visual>
      <collision>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Outer_Link.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
    </collision>
  </link>

  <link name="${prefix}hand">
    <visual>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Hand.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="white"/>
    </visual>
      <collision>
      <geometry>
        <mesh filename="package://pf400_description/meshes/PF400_Hand.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <origin xyz="0 0 0.074" rpy="0 0 0"/>
    </collision>
  </link>

  <link name="${prefix}left_finger">
    <visual>
      <origin xyz="-0.02 -0.053 0.053" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://pf400_description/meshes/pf400_left_finger.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="aluminum"/>
    </visual>
      <collision>
      <geometry>
        <mesh filename="package://pf400_description/meshes/pf400_left_finger.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <origin xyz="-0.02 -0.053 0.053" rpy="0 0 0"/>
    </collision>
  </link>

  <link name="${prefix}right_finger">
    <visual>
      <origin xyz="-0.021 -0.028 0.105" rpy="-1.5708 0 0"/>
      <geometry>
        <mesh filename="package://pf400_description/meshes/pf400_right_finger.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="aluminum"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://pf400_description/meshes/pf400_right_finger.STL" scale="0.001 0.001 0.001"/>
      </geometry>
      <origin xyz="-0.021 -0.028 0.105" rpy="-1.5708 0 0"/>
    </collision>
  </link>

  <!-- Joints -->

  <joint name="J1" type="prismatic">
    <parent link="${prefix}z_column"/>
    <child link="${prefix}z_carriage"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0.03005 -0.0198 0.17535" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0.002" upper="1.160" velocity="0.5"/> 
  </joint>

  <joint name="J2" type="revolute">
    <parent link="${prefix}z_carriage"/>
    <child link="${prefix}inner_link"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0.173001 0.020 0.0235" rpy="0 0 0"/>
    <limit effort="1000.0" lower="-1.62316" upper="1.62316" velocity="0.5"/>
  </joint>

  <joint name="J3" type="revolute">
    <parent link="${prefix}inner_link"/>
    <child link="${prefix}outer_link"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0.302 0 -0.072285" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0.175" upper="6.1" velocity="0.5"/>
  </joint>

  <joint name="J4" type="revolute">
    <parent link="${prefix}outer_link"/>
    <child link="${prefix}hand"/>
    <axis xyz="0 0 1"/>
    <origin xyz="0.289 0 -0.032942" rpy="0 0 0"/>
    <limit effort="1000.0" lower="-16.7552" upper="16.7552" velocity="0.5"/>
  </joint>

  <joint name="J5" type="prismatic">
    <parent link="${prefix}hand"/>
    <child link="${prefix}left_finger"/>
    <axis xyz="0 1 0"/>
    <origin xyz="0.078573 0 -0.04269" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0.04" upper="0.0675" velocity="0.5"/>
  </joint>

  <joint name="J5_mirror" type="prismatic">
    <parent link="${prefix}hand"/>
    <child link="${prefix}right_finger"/>
    <axis xyz="0 -1 0"/>
    <origin xyz="0.078573 0 -0.04269" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0.04" upper="0.0675" velocity="0.5"/>
    <mimic joint="J5" multiplier="1"/>
  </joint>

  <joint name="J6" type="prismatic">
    <parent link="${prefix}rail_base"/>
    <child link="${prefix}rail"/>
    <axis xyz="1 0 0"/>
    <origin xyz="0 0 0.032" rpy="0 0 0"/>
    <limit effort="1000.0" lower="-1" upper="1" velocity="0.5"/>
  </joint>

  <joint name="rail_link" type="fixed">
    <parent link="${prefix}rail"/>
    <child link="${prefix}z_column"/>
    <axis xyz="0 1 0"/>
    <origin xyz="-0.1005 0 0.0157" rpy="0 0 0"/>
    <limit effort="1000.0" lower="0" upper="0" velocity="0.5"/>
  </joint>
  </xacro:macro>

</robot>
